src/main/java/be/seeseemelk/mockbukkit/entity/AllayMock.java

```diff
@@ -150,78 +155,79 @@ public void assertInventoryContains(ItemStack item, String s)
 	@Override
 	public boolean canDuplicate()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.canDulicate;
 	}
 
 	@Override
 	public void setCanDuplicate(boolean canDuplicate)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.canDulicate = canDuplicate;
 	}
 
 	@Override
 	public long getDuplicationCooldown()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.duplicationCoolDown;
 	}
 
 	@Override
 	public void setDuplicationCooldown(long cooldown)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.duplicationCoolDown = cooldown;
 	}
 
 	@Override
 	public void resetDuplicationCooldown()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		this.duplicationCoolDown = 6000L;
+		this.canDulicate = false;
 	}
 
 	@Override
 	public boolean isDancing()
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.isDancing;
 	}
 
 	@Override
 	public void startDancing(@NotNull Location location)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkArgument(location != null, "Location cannot be null");
+		Preconditions.checkArgument(location.getBlock().getType() == Material.JUKEBOX,
+				"Location must be a Jukebox");
+
+		this.isDancing = true;
+		this.jukebox = location;
```

Comment:
```suggestion
		this.jukebox = location.clone();
```
Some weird behavior could come from using the same reference as provided externally. There's for example the Location#add() method which can modify the object externally after the location has been set.